src/Microsoft.AspNetCore.Routing.Abstractions/RouteValueDictionary.cs

```diff
@@ -16,14 +17,18 @@ namespace Microsoft.AspNetCore.Routing
     /// </summary>
     public class RouteValueDictionary : IDictionary<string, object>, IReadOnlyDictionary<string, object>
     {
-        internal Storage _storage;
+        private const int DefaultCapacity = 4;
+
+        internal KeyValuePair<string, object>[] _arrayStorage;
+        internal PropertyStorage _propertyStorage;
+        private int _count;
 
         /// <summary>
         /// Creates an empty <see cref="RouteValueDictionary"/>.
         /// </summary>
         public RouteValueDictionary()
```

Comment:
Wondering if having one more constructor which takes in an 'initial capacity' would help here. For example, DfaMatcher knows the number of parameters to process and add to a RVD

Reply:
I considered doing that but that number would generally be 4 or less, so it kinda evens out.

Reply:
I guess what I'm trying to say is that I'd rather let this decision be motivated by data rather than speculation. Based on the current tests there's no justification for adding a capacity.

Reply:
Sure, sounds good 👍